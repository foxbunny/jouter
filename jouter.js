const TOKEN_RE = /:[^\/]+/g
const ANY_RE = /\*/g
const SUBPATH_RE = /\/\.\.\./g

const regexify = x =>
  x.replace(TOKEN_RE, '([^/]+)')
   .replace(ANY_RE, '.*')
   .replace(SUBPATH_RE, '(\/.*)')

// parseRoute :: String | RegExp -> RegExp
const routeRe = x =>
  x instanceof RegExp ?
    new RegExp(x.source)
  : new RegExp(`^${regexify(x)}$`)

// route ::  (* -> *), String -> (String -> _)
const route = (f, r) => {
  const re = routeRe(r)
  return (path) => {
    const match = re.exec(path)
    if (!match) return
    f(...match.slice(1))
    return true
  }
}

// Unsafe functions
const pathHandler = {
  // get :: _ -> String
  get: () => window.location.pathname,

  // set :: (String, String) -> _
  set: (path, title) =>
    window.history.pushState(undefined, title, path),

  // listen :: (* -> *) -> _
  listen: f => window.onpopstate = f,

  // decorate :: (* -> *) -> (* -> *)
  decorate: f => f,

  // onNoMatch :: (String -> _) -> _
  onNoMatch: f => undefined,
}

export const createRouter = (myPathHandler = {}) => {
  let routes = []
  const path = {
    ...pathHandler,
    ...myPathHandler
  }

  const add = (f, r) => 
    routes.push(route(path.decorate(f), r))

  const dispatch = p => {
    const hadMatch = routes.reduce((result, f) => {
      const match = f(p)
      return result || match
    }, false) 
    if (!hadMatch) path.onNoMatch(p)
    return hadMatch
  }

  const go = (p, t) => {
    path.set(p, t)
    dispatch(path.get())
  }

  const handleEvent = e => {
    e.preventDefault()
    const target = e.currentTarget
    go(target.href, target.title)
  }

  const start = () => {
    path.listen(() => dispatch(path.get()))
    dispatch(path.get())
  }

  const dispatchRoutes = subpath =>
    dispatch(subpath)

  dispatchRoutes.add = add
  dispatchRoutes.go = go
  dispatchRoutes.handleEvent = handleEvent
  dispatchRoutes.start = start

  return dispatchRoutes
}

export default createRouter
