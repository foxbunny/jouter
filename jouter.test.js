import {routeRe, route, createRouter} from './jouter'

describe('createRouter', () => {
  let currentPath = '/foo'
  let listeners = []

  beforeEach(() => {
    currentPath = '/foo'
    listeners = []
  })

  const fakePathHandler = {
    get: jest.fn(() => currentPath),
    set: jest.fn((path, title) => {
      currentPath = path
      listeners.forEach(f => f())
    }),
    listen: jest.fn(f => listeners.push(f))
  }

  test('will return a router object', () => {
    const router = createRouter()
    expect(router.add).toBeInstanceOf(Function)
    expect(router.start).toBeInstanceOf(Function)
    expect(router.handleEvent).toBeInstanceOf(Function)
    expect(router.go).toBeInstanceOf(Function)
  })

  test('switching paths', () => {
    const router = createRouter(fakePathHandler)
    router.go('/test')
    expect(currentPath).toBe('/test')
  })

  test('adding and handling routes', () => {
    const router = createRouter(fakePathHandler)
    const fn = jest.fn()
    router.add(fn, '/:x')
    router.go('/bar')
    expect(fn).toHaveBeenCalledWith('bar')
  })

  test('start routing', () => {
    const router = createRouter(fakePathHandler)
    const fn = jest.fn()
    router.add(fn, '/:x')
    router.start()
    expect(listeners.length).toBe(1)
    listeners[0]()
    expect(fn).toHaveBeenCalledWith('foo')
  })

  test('handing events', () => {
    const router = createRouter(fakePathHandler)
    const fn = jest.fn()
    const elem = document.createElement('a')
    elem.href = '/baz'
    router.add(fn, '/:x')
    elem.addEventListener('click', router.handleEvent)
    elem.dispatchEvent(new Event('click'));
    expect(fn).toHaveBeenCalledWith('baz')
    expect(currentPath).toBe('/baz')
  })

  test('router object is a function', () => {
    const router = createRouter(fakePathHandler)
    expect(router).toBeInstanceOf(Function)
  })

  test('router objects can be used as route handlers', () => {
    const router = createRouter(fakePathHandler)
    const subrouter = createRouter(fakePathHandler)
    const fn = jest.fn()
    router.add(subrouter, '/sub/...')
    subrouter.add(fn, '/foo/:x')
    router.go('/sub/foo/2')
    expect(fn).toHaveBeenCalledWith('2')
  })

  test('router object can be created with a decorator function', () => {
    const injectedDependency = {foo: 'bar'}
    const fn = jest.fn()
    const router = createRouter({
      ...fakePathHandler, 
      decorate: fn => (...args) => fn(injectedDependency, ...args)
    })
    router.add(fn, '/:x')
    router.go('/12')
    expect(fn).toHaveBeenCalledWith(injectedDependency, '12')
  })

  test('unmatched route will invoke the error handler', () => {
    const errorHandler = jest.fn()
    const router = createRouter({
      ...fakePathHandler,
      onNoMatch: errorHandler,
    })
    router.add(jest.fn(), '/:x')
    router.go('/')
    expect(errorHandler).toHaveBeenCalledWith("/")
  });

  test('error handler is not invoked if subroute matches', () => {
    const errorHandler = jest.fn();
    const router = createRouter({
      ...fakePathHandler,
      onNoMatch: errorHandler
    })
    const subrouter = createRouter(fakePathHandler)
    router.add(subrouter, '/sub/...')
    subrouter.add(jest.fn(), '/foo/:x')
    router.go('/sub/foo/2')
    expect(errorHandler).not.toHaveBeenCalled()
  });

  test('if subroute is unmatched, error handler is invoked', () => {
    const errorHandler = jest.fn();
    const router = createRouter({
      ...fakePathHandler,
      onNoMatch: errorHandler
    })
    const subrouter = createRouter(fakePathHandler)
    router.add(subrouter, '/sub/...')
    subrouter.add(jest.fn(), '/foo/:x')
    router.go('/bogus')
    expect(errorHandler).toHaveBeenCalledWith('/bogus')
  })
})
