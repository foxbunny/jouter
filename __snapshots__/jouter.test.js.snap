exports[`routeRe will convert tokens to regexp 1`] = `/^\/([^\/]+)\/drink-([^\/]+)$/`;

exports[`routeRe will match prefix wildcards 1`] = `/^\/foo(\/.*)$/`;

exports[`routeRe will match route as is if no tokens 1`] = `/^\/about$/`;

exports[`routeRe will match wildcard 1`] = `/^\/foo\/.*$/`;
